### PR TITLE
Secure Services with beforeResolver

### LINKED ISSUE #1630: RFC - Secure Services by default
As discussed in the 2021-11-12 Core Meeting, we want to do what we can to make services secure by default. Currently, if you don't add something like `requireAuth()` manually in a services function, any GraphQL endpoints that resolve to those services will happily execute by anyone on the internet. That could be bad.

This issue will serve as a single place to discuss different ways to make services secure by default. Here "by default" means that if you did not do anything special to explicitly allow access, your service/endpoint would be secure—return an error when attempting to access with no authorization/authentication provided.

## Current State

Given the following service function:

```javascript
export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

And the following GraphQL SDL:

```graphql
export const schema = gql`
  type User {
    id: Int!
    name: String!
    email: String!
  }

  type Mutation {
    deleteUser(id: Int!): User!
  }
```

Anyone on the internet could perform this mutation and delete every user in the database:

```javascript
import { request, gql } from 'graphql-request'

const query = gql`
  mutation DeleteUser($id: Int!) {
    deleteUser(id: $id) {
      id
    }
  }
`

for (let i=0; i<100000; i++) {
  try {
    await request('https://insecuresite.com/.netlify/functions/graphql', query, { id: i })
  } catch (e) {
    console.log(`User ${i} not found, trying next...`)
  }
}
```

Note that a service can be used internally as well, and the same security implications apply.

This can be prevented by simply adding a `requireAuth()` call inside the service to check for proper permission (either being logged in at all, or the user having a specific role):

```javascript
export const deleteUser = ({ id }) => {
  requireAuth({ roles: ['admin'] })

  return db.user.delete({
    where: { id },
  })
}
```

However, having to add it manually is rife for human error. We want Redwood to be secure by default and help save users from themselves.

## Proposal A

Use some form of `requireAuth()` behind the scenes and add a new function like `allowAccess()` to explicitly allow access.

Assuming the same service:

```javascript
export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

Any access to this service would raise an error. You would need to explicitly allow access:

```javascript
export const deleteUser = ({ id }) => {
  allowAccess({ roles: ['admin'] })
  return db.user.delete({
    where: { id },
  })
}
```

Perhaps just calling `allowAccess()` (without arguments) allows any logged in user to access.

Some downsides to this proposal:

* Before adding `allowAccess()` calls, it's not immediately clear that any kind of authorization is happening at all, these just appear to be regular functions
* Not backwards compatible with existing codebases

## Proposal B

This proposal addresses some of the concerns in Proposal A. This solution would be backwards compatible (since it uses the same `requireAuth()` function we already know and love). We add a convention of exporting a function that will automatically run before all services so you only need to declare `requireAuth()` once:

```javascript
export const beforeService = () => {
  requireAuth({ roles: ['admin'] })
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

If you have multiple service functions and want to not require auth for one or more of them:

```javascript
export const beforeService = () => {
  requireAuth({ roles: ['admin'] }, except: ['listUsers'])
}

export const listUsers = () => {
  return db.user.findMany()
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

Or the opposite where you only want to lock down a few functions:

```javascript
export const beforeService = () => {
  requireAuth({ roles: ['admin'] }, only: ['deleteUser'])
}

export const listUsers = () => {
  return db.user.findMany()
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

This proposal has the added benefit of letting you execute additional code that you always want to occur before a group of services (logging the API call, perhaps):

```javascript
import { logAccess } from 'src/services/logger'

export const beforeService = ({ context, serviceName }) => {
  requireAuth({ roles: ['admin'] }, only: ['deleteUser'])
  logAccess(context.currentUser, serviceName)
}

export const deleteUser = ({ id }) => {
  return db.user.delete({
    where: { id },
  })
}
```

What if the service does not export a `beforeService` function? We have a few options:

* Raise an error: if you decide you do not need authorization in your service then you need to make it an explicit, conscious decision by exporting a `beforeService()` that returns nothing
* Not require a `beforeService()` call, but throw an error or at least a warning if we do not encounter a `requireAuth()` function call in a service
* Require an explicit function call like `skipAuth()` call to really make it clear what you're doing

## Conclusion

Please add a comment if you have any additional syntax ideas!

### PR DESCRIPTION
**Pull request by cannikin**
Closes #1630

### CODE REVIEW

File: packages/cli/src/commands/setup/auth/auth.js [hunk 1]
```diff
@@ -234,8 +238,30 @@ export const builder = (yargs) => {
...
+  console.log(`✋ ~ file: auth.js ~ line 238 ~ handler ~ provider`, provider)
+  console.log(`✋ ~ file: auth.js ~ line 238 ~ handler ~ force`, force)
   const providerData = await import(`./providers/${provider}`)
 
+  // check if api/src/lib/auth.js already exists and if so, ask the user to overwrite
+  if (force === false) {
+    console.log(
+      `✋ ~ file: auth.js ~ line 247 ~ handler ~ Object.keys(files(provider))[0])`,
+      Object.keys(files(provider))[0]
+    )
```
Comment (peterp) [inline packages/cli/src/commands/setup/auth/auth.js:250]: 👋

File: packages/api/src/makeServices.ts [hunk 2]
```diff
@@ -1,9 +1,46 @@
...
+      'NOTICE: Redwood v1.0 will make services secure by default. To optin to this behavior now, add `experimentalSecureServices = true` to the [api] section of your redwood.toml file. For more information: https://redwoodjs.com/docs/secure-services'
+    )
+    return services
+  }
+
+  const servicesCollection: ServicesCollection = {}
+
+  for (const [name, resolvers] of Object.entries(services)) {
+    if (!resolvers?.beforeResolver) {
+      throw new MissingBeforeResolver(name)
```
Comment (peterp) [inline packages/api/src/makeServices.ts:18]: I think it's helpful to append Error to Errors :)
```suggestion
      throw new MissingBeforeResolverError(name)
```

Comment (cannikin) [reply packages/api/src/makeServices.ts]: Fixed!

**Review by peterp — APPROVED** [review]
This looks good to me - I would like to get it merged early so that I can start testing it in canary for a few days before we release it to the public

**Pull request merged**